pkg/cmd/stack/wait.go

```diff
@@ -0,0 +1,191 @@
+	isCompose, err := strconv.ParseBool(configMap.Data[ComposeField])
+	if err != nil {
+		return nil, err
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(configMap.Data[YamlField])
+	if err != nil {
+		return nil, err
+	}
+
+	stack, err := model.ReadStack(decoded, isCompose)
+	if err != nil {
+		return nil, err
+	}
+	stack.Name = configMap.Data[NameField]
+	stack.Namespace = configMap.Namespace
+	return stack, nil
+}
+
+func addHiddenExposedPortsToDependentSvcs(ctx context.Context, s *model.Stack, svc *model.Service) {
+	for dependentSvcName := range svc.DependsOn {
+		addHiddenExposedPortsToSvc(ctx, s.Services[dependentSvcName], s.Namespace)
+	}
+}
+
+func waitForSvcs(ctx context.Context, stack *model.Stack, svc *model.Service, c kubernetes.Interface) error {
+	var err error
+	for dependentSvc, condition := range svc.DependsOn {
+		fmt.Printf("Waiting for service '%s'\n", dependentSvc)
+		switch condition.Condition {
+		case model.DependsOnServiceRunning:
+			err = waitForSvcRunning(ctx, stack, dependentSvc, c)
+			if err != nil {
+				return err
+			}
+		case model.DependsOnServiceHealthy:
+			err = waitForSvcHealthy(stack, dependentSvc)
+			if err != nil {
+				return err
+			}
+		case model.DependsOnServiceCompleted:
+			err = waitForSvcCompleted(ctx, stack, dependentSvc, c)
+			if err != nil {
+				return err
+			}
+		}
+		fmt.Printf("Service '%s' is ready\n", dependentSvc)
+	}
+	return nil
+}
+
+func waitForSvcRunning(ctx context.Context, stack *model.Stack, svcName string, c kubernetes.Interface) error {
+	svcToWaitFor := stack.Services[svcName]
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	timeout := time.Now().Add(300 * time.Second)
+	isDeployment := isDeployment(svcToWaitFor)
+	isStatefulset := isStatefulset(svcToWaitFor)
+
+	for time.Now().Before(timeout) {
+		<-ticker.C
+		switch {
+		case isDeployment:
+			if deployments.IsRunning(ctx, stack.Namespace, svcName, c) {
+				return nil
+			}
+		case isStatefulset:
+			if statefulsets.IsRunning(ctx, stack.Namespace, svcName, c) {
+				return nil
+			}
+		default:
+			if jobs.IsRunning(ctx, stack.Namespace, svcName, c) {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("Service '%s' is taking too long to be healthy. Please check logs and try again later.", svcName)
+}
+
+func isDeployment(svc *model.Service) bool {
+	return len(svc.Volumes) == 0 && svc.RestartPolicy == apiv1.RestartPolicyAlways
+}
+func isStatefulset(svc *model.Service) bool {
+	return len(svc.Volumes) != 0 && svc.RestartPolicy == apiv1.RestartPolicyAlways
+}
+
+func waitForSvcHealthy(stack *model.Stack, svcName string) error {
+	svcToWaitFor := stack.Services[svcName]
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	timeout := time.Now().Add(300 * time.Second)
+
+	for time.Now().Before(timeout) {
+		<-ticker.C
+		for _, p := range svcToWaitFor.Ports {
+			url := fmt.Sprintf("http://%s:%d/", svcName, p.ContainerPort)
+			resp, err := http.Get(url)
```

Comment:
is this how `wait-for` is implemented? what if it is not an HTTP port?

Reply:
I can make a callback to https and ftp, but I think it could be very useful if we support docker-compose `healthcheck` field which is the command that must be ran to know if the  service is healthy instead of checking all the ports with all the protocols.